### PR TITLE
DD-7: delete contract interactions, add pallet calls

### PR DESCRIPTION
**Pull request by lana-shanghai**
now it is possible to send each sim transaction to our Canvas node!

### CODE REVIEW

File: requirements/base.txt [hunk 1]
```diff
@@ -2,30 +2,43 @@
...
 sortedcontainers==2.3.0   # via -r requirements/base.in
+substrate-interface==0.11.10  # via -r requirements/base.in
 unidecode==0.4.21         # via awesome-slugify
+urllib3==1.26.2           # via requests
+websocket-client==0.57.0  # via substrate-interface
+xxhash==2.0.0             # via substrate-interface
 
 # The following packages are considered to be unsafe in a requirements file:
-pip==20.3.1               # via -r requirements/base.in, pip-tools
+# pip
```
Comment (spyrostz) [inline requirements/base.txt:44]: Did `fab compile` produce that ? If yes no problem there, if you did this manually I think that it has to be reverted.

File: requirements/tests.in [hunk 2]
```diff
@@ -11,3 +11,4 @@ pytest
 ptpython
 requests-mock
 deepdiff
+substrate-interface
```
Comment (spyrostz) [inline requirements/tests.in:14]: No need to have it here. tests.in include all the dependencies of base.in, thus substrate-interface should be included.

File: src/d3a/models/market/blockchain_interface.py [hunk 3]
```diff
@@ -33,12 +31,12 @@
     from d3a.models.market.blockchain_utils import create_market_contract, create_new_offer, \
         cancel_offer, trade_offer
     if ENABLE_SUBSTRATE is True:
```
Comment (faizan2590) [inline src/d3a/models/market/blockchain_interface.py:33]:
```suggestion
    if ENABLE_SUBSTRATE:
```

File: src/d3a/models/market/blockchain_interface.py [hunk 4]
```diff
@@ -199,12 +97,21 @@ def handle_blockchain_trade_event(self, offer, buyer, original_offer, residual_o
...
-        path = Path.cwd().joinpath('src', 'd3a', 'models', 'market', 'metadata.json')
-        trade_id, call = self.storage_contract_call(
-            self.contracts['main'],
-            "trade",
-            str(path)
+
+        call_params = {
+            'id': 3,
+            'buyer': ALICE_STASH_ADDRESS,
+            'seller': BOB_STASH_ADDRESS,
```
Comment (faizan2590) [inline src/d3a/models/market/blockchain_interface.py:104]: Ideally all agents should be having their own unique addresses unlike here where all agents are sharing the same address

File: src/d3a/models/market/blockchain_interface.py [hunk 5]
```diff
@@ -199,12 +97,21 @@ def handle_blockchain_trade_event(self, offer, buyer, original_offer, residual_o
...
+            'amount': default_amount,
+            'rate': default_rate
```
Comment (faizan2590) [inline src/d3a/models/market/blockchain_interface.py:106]:
```suggestion
            'rate': int((trade.offer.price/trade.offer.energy))
```

File: src/d3a/models/market/blockchain_interface.py [hunk 6]
```diff
@@ -199,12 +97,21 @@ def handle_blockchain_trade_event(self, offer, buyer, original_offer, residual_o
...
-        trade_id, call = self.storage_contract_call(
-            self.contracts['main'],
-            "trade",
-            str(path)
+
+        call_params = {
+            'id': 3,
+            'buyer': ALICE_STASH_ADDRESS,
+            'seller': BOB_STASH_ADDRESS,
+            'amount': default_amount,
```
Comment (faizan2590) [inline src/d3a/models/market/blockchain_interface.py:105]: Its not helpful to always document trade with a default_amount. I would rather vouch for to document an actual traded energy here

File: src/d3a/models/market/blockchain_utils.py [hunk 7]
```diff
@@ -28,16 +25,20 @@
 log = getLogger(__name__)
 
 
-BC_NUM_FACTOR = 10 ** 10
-BOB_ADDRESS = "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty"
-ALICE_ADDRESS = "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY"
+BC_NUM_FACTOR = 10 ** 12
+BOB_STASH_ADDRESS = "5HpG9w8EBLe5XCrbczpwq5TSXvedjrBGCwqxK1iQ7qUsSWFc"
+ALICE_STASH_ADDRESS = "5GNJqTPyNqANBkUVMN1LPPrxXnFouWXoe2wNSmmEoLctxiZY"
```
Comment (faizan2590) [inline src/d3a/models/market/blockchain_utils.py:30]: May be add a module `blockhain.constants.py` and move all associated constants there

**Review by faizan2590 — COMMENTED** [review]
Left some comments, will give another look once addressed.

Thanks for integrating substrate to d3a!

File: requirements/base.txt (hunk 1 above)
Comment (lana-shanghai) [reply requirements/base.txt:44]: Hey Spyros, I didn't do it manually

File: requirements/tests.in (hunk 2 above)
Comment (lana-shanghai) [reply requirements/tests.in:14]: I think it was failing for me for some reason, but I will test to see if it's OK without it

File: src/d3a/models/market/blockchain_interface.py (hunk 3 above)
Comment (lana-shanghai) [reply src/d3a/models/market/blockchain_interface.py:33]: Right!

File: src/d3a/models/market/blockchain_utils.py (hunk 7 above)
Comment (lana-shanghai) [reply src/d3a/models/market/blockchain_utils.py:30]: Makes sense, I will bring them out

File: src/d3a/models/market/blockchain_interface.py (hunk 6 above)
Comment (lana-shanghai) [reply src/d3a/models/market/blockchain_interface.py:105]: For the sake of simplicity, I did all defaults to make sure the blockchain stuff works. Should we do this in another ticket?

Comment (faizan2590) [reply src/d3a/models/market/blockchain_interface.py:105]: sounds good, please add a ticket to deal with it later.

File: src/d3a/models/market/blockchain_interface.py (hunk 4 above)
Comment (lana-shanghai) [reply src/d3a/models/market/blockchain_interface.py:104]: This is part of the DID epic and I am not sure we should do it in this ticket... Or do you suggest generating a random address for every participant?

Comment (faizan2590) [reply src/d3a/models/market/blockchain_interface.py:104]: Hmm good idea, better post a link of this comment to DID epic in order for the first thing that needs to be changed while implementing DID

Comment (lana-shanghai) [reply src/d3a/models/market/blockchain_interface.py:104]: Added https://gridsingularity.atlassian.net/browse/DD-12

File: src/d3a/models/market/blockchain_utils.py (hunk 7 above)
Comment (lana-shanghai) [reply src/d3a/models/market/blockchain_utils.py]: Added commit https://github.com/gridsingularity/d3a/pull/1033/commits/13ba9779d19c47c0ec2e7ce9816df5f209e62f21

File: src/d3a/models/market/blockchain_interface.py (hunk 6 above)
Comment (lana-shanghai) [reply src/d3a/models/market/blockchain_interface.py:105]: https://gridsingularity.atlassian.net/secure/RapidBoard.jspa?rapidView=43&projectKey=DD&view=detail&selectedIssue=DD-14

File: src/d3a/models/market/blockchain_interface.py (hunk 5 above)
Comment (lana-shanghai) [reply src/d3a/models/market/blockchain_interface.py:106]: Added suggestion to ticket DD-14

File: requirements/tests.in (hunk 2 above)
Comment (lana-shanghai) [reply requirements/tests.in]: Done

**Review by faizan2590 — COMMENTED** [review]
LGTM

File: src/d3a/models/market/blockchain_interface.py [hunk 8]
```diff
@@ -99,7 +99,6 @@ def handle_blockchain_trade_event(self, offer, buyer, original_offer, residual_o
     def track_trade_event(self, trade):
 
         call_params = {
-            'id': 3,
```
Comment (faizan2590) [inline src/d3a/models/market/blockchain_interface.py]: As mentioned in other thread in similar pr(https://github.com/gridsingularity/canvas-node/pull/6#discussion_r565466825), I would be in favour of using trade id here which is already unique.

File: src/d3a/blockchain/constants.py [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+default_call_module = 'Tradestorage'
+default_call_function = 'store_trade_map'
+default_amount = {
+    'free': 1 * BC_NUM_FACTOR,
+    'reserved': 0,
+    'miscFrozen': 0,
+    'feeFrozen': 0
+}
+default_rate = 12
+address_type = 42
```
Comment (faizan2590) [inline src/d3a/blockchain/constants.py:16]: minor but can you also move flag `ENABLE_SUBSTRATE` from `src/d3a/blockchain/__init__.py` to here in order to have common place for all constants

**Review by faizan2590 — DISMISSED** [review]
One nice to have comment, other than that LGTM

File: src/d3a/models/market/blockchain_interface.py [hunk 10]
```diff
@@ -71,121 +69,21 @@ def bc_listener(self):
         pass
 
 
-class SubstrateBlockchainInterface:
-    def __init__(self, bc):
-        self.contracts = {'main': main_address}
-        self.substrate = bc
+class SubstrateBlockchainInterface(BlockChainInterface):
+    def __init__(self):
+        super().__init__()
```
Comment (spyrostz) [inline src/d3a/models/market/blockchain_interface.py:74]: I think you can completely omit the constructor, this is the default one.

**Review by spyrostz — APPROVED** [review]
One minor / optional comment, LGTM

**Review by faizan2590 — APPROVED** [review]

File: src/d3a/models/market/blockchain_interface.py (hunk 10 above)
Comment (lana-shanghai) [reply src/d3a/models/market/blockchain_interface.py:74]: I will add to the blockchain ticket that deals with the offer functions in blockchain utils

**Pull request merged**